pints/_mcmc/_slice_stepout.py

```diff
@@ -0,0 +1,430 @@
+    reduces the size of the interval after rejecting a trial point,
+    as defined in [1] Fig. 5. pp.716. This algorithm consists of the
+    following steps:
+
+    1. ``\bar{L} = L`` and ``\bar{R} = R``
+    2. Repeat:
+        a. ``U \sim uniform(0, 1)``
+        b. ``x_1 = \bar{L} + U (\bar{R} - \bar{L})``
+        c. if ``y < f(x_1)`` accept ``x_1`` and exit loop,
+           else
+            if ``x_1 < x_0``, ``\bar{L} = x_1``
+            else ``\bar{R} = x_1``
+
+    Intuitively, we uniformly sample a trial point from the interval ``I``,
+    and subsequently shrink the interval each time a trial point is rejected.
+
+    To avoid floating-point underflow, we implement the suggestion advanced
+    in [1] pp.712. We use the log pdf of the un-normalised posterior
+    (``g(x) = log(f(x))``) instead of ``f(x)``. In doing so, we use an
+    auxiliary variable ``z = log(y) = g(x0) − \epsilon``, where
+    ``\epsilon \sim \text{exp}(1)`` and define the slice as
+    S = {x : z < g(x)}.
+
+    [1] Neal, R.M., 2003. Slice sampling. The annals of statistics, 31(3),
+    pp.705-767.
+    """
+
+    def __init__(self, x0, sigma0=None):
+        super(SliceStepoutMCMC, self).__init__(x0, sigma0)
+
+        # Set initial state
+        self._x0 = np.asarray(x0, dtype=float)
+        self._running = False
+        self._ready_for_tell = False
+
+        # Iterations monitoring
+        self._mcmc_iteration = 0
+
+        # Current sample, log_pdf of the current sample
+        self._current = None
+        self._current_log_pdf = None
+
+        # Current log_y used to define the slice
+        self._current_log_y = None
+
+        # Current proposed sample
+        self._proposed = None
+
+        # Default initial interval width w used in the Stepout procedure
+        # to expand the interval
+        self._w = np.ones(len(self._x0))
+
+        # Default integer limiting the size of the interval to ``m*w```
+        self._m = 50
+
+        # Flag to initialise the expansion of the interval ``I=(L,R)``
+        self._first_expansion = False
+
+        # Flag indicating whether the interval expansion is concluded
+        self._interval_found = False
+
+        # Number of steps used for expanding the interval ``I=(L,R)``
+        self._j = None
+        self._k = None
+
+        # Edges of the interval ``I=(L,R)``
+        self._l = None
+        self._r = None
+
+        # Multi-dimensional points used to calculate the log_pdf of the edges ``l,r```
+        self._temp_l = None
+        self._temp_r = None
+
+        # Log_pdf of interval edge points ``l,r```
+        self._fx_l = None
+        self._fx_r = None
+
+        # Flags to indicate the interval edge to update
+        self._set_l = False
+        self._set_r  = False
+
+        # Index of parameter "xi" we are updating of the sample "x = (x1,...,xn)"
+        self._i = 0
+
+        # Stochastic variables as instance variables for testing
```

Comment:
Not sure what an instance variable is. Think the variables could just be commented as intermediate variables used in slice sampling and stepout procedures.